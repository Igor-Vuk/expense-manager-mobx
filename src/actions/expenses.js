import { action } from 'mobx'
import RootStore from '../store/RootStore'
import uuid from 'uuid'

/* With decorators */
class ExpensesAction {
  
  @action addExpense = (expense) => {
    const newExpense = this.addExpenseAction(expense)
    RootStore.ExpensesStore.expenses.push(newExpense)
  }

  @action removeExpense = (id) => {
    const filterExpense = this.removeExpenseAction(id)
    RootStore.ExpensesStore.expenses = filterExpense
  }

  @action editExpense = (id, updates) => {
    const editExpense = this.editExpenseAction(id, updates)
    RootStore.ExpensesStore.expenses = editExpense
  }

  /* Add expense action */
  addExpenseAction = ({
    description = '',
    note = '',
    amount = 0,
    createdAt = 0
  } = {}) => {
    return {
      id: uuid(),
      description,
      note,
      amount,
      createdAt
    }
  }

  /* Remove expense action */
  removeExpenseAction = ({
    id=''
  } = {}) => {
    return RootStore.ExpensesStore.expenses.filter((expense) => expense.id !== id )
  }

  /* Edit expense action */
  editExpenseAction = (id, updates) => {
    return RootStore.ExpensesStore.expenses.map((expense) => {
      if(expense.id === id) {
        return {
          ...expense,
          ...updates
        }
      } else {
        return {
          ...expense
        }
      }
    })
  }

}

const actions = new ExpensesAction()
export default actions


// Wite addExpense action as just one function

// @action addExpense = ({
//   description = '',
//   note = '',
//   amount = 0,
//   createdAt = 0
// } = {}) => {
//   const expense = {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
//   RootStore.ExpensesStore.expenses.push(expense)
// }


/* Without decorators */
// const addExpense = action(({
//   description = '',
//   note = '',
//   amount = 0,
//   createdAt = 0
// } = {}) => {
//   const expense = {
//     id: uuid(),
//     description,
//     note,
//     amount,
//     createdAt
//   }
//   ExpensesStore.expenses.push(expense)
// })

// export {addExpense} 

