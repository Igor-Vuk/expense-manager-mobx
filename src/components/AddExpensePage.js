import React, {Component} from 'react'
import { inject } from 'mobx-react'
import ExpenseForm from './ExpenseForm'

/* With decorators */
@inject('ExpensesAction')
class AddExpensePage extends Component {

  onSubmit = (expense) => {
    this.props.ExpensesAction.addExpense(expense)
    this.props.history.push('/')
  }

  render() {
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm
          onSubmit={this.onSubmit}
        />
      </div>
    )
  }
}

/* Without decorators */
// const AddExpensePage = inject('ExpensesAction')((props) => (
//   <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm
//       onSubmit={(expense) => {
//         props.ExpensesAction.addExpense(expense)
//         props.history.push('/')
//       }}
//     />
//   </div>
// ))

export default AddExpensePage
