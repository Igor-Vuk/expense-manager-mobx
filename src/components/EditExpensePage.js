import React, { Component } from 'react'
import { inject } from 'mobx-react'
import ExpenseForm from './ExpenseForm'

@inject('RootStore', 'ExpensesAction')
class EditExpensePage extends Component {

  editExpense = (expense) => {
    this.props.ExpensesAction.editExpense(expense.id, expense)
    this.props.history.push('/')
  }

  removeExpense = (findExpense) => {
    this.props.ExpensesAction.removeExpense({id: findExpense.id})
    this.props.history.push('/')
  } 
  
  render() {
    const findExpense = this.props.RootStore.ExpensesStore.findExpense(this.props.match.params.id)
    return (
      <div>
        <ExpenseForm
          expense={findExpense}
          onSubmit={this.editExpense}
        />
        <button 
          onClick={() => {this.removeExpense(findExpense)}}>
          Remove item
        </button>
      </div>
    )
  }
}

export default EditExpensePage