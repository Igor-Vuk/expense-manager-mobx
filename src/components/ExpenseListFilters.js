import React, {Component} from 'react'
import { observer, inject } from 'mobx-react'
import 'react-dates/initialize'
import { DateRangePicker } from 'react-dates'

/* With decorators */
@inject('RootStore', 'FiltersAction')
@observer
class ExpenseListFilters extends Component {
  state = {
    calendarFocused: null
  }
  
  onDatesChange = ({startDate, endDate}) => {
    this.props.FiltersAction.setStartDate(startDate)
    this.props.FiltersAction.setEndDate(endDate)
  }

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({
      calendarFocused
    }))
  }

  onTextChange = (e) => {
    this.props.FiltersAction.setTextFilter(e.target.value)
  }

  onSortChange = (e) => {
    if(e.target.value === "date") {
      this.props.FiltersAction.sortByDate()
    } else if (e.target.value === "amount") {
      this.props.FiltersAction.sortByAmount()
    }
  }

  render() {
    const {text, sortBy} = this.props.RootStore.FiltersStore.filters
    return (
      <div>

        <div>
          <input type="text" value={text} onChange={this.onTextChange} placeholder="Search for expense"/>
        </div>

        <br/>

        <div>
          <select
            value={sortBy}
            onChange={this.onSortChange}>
            <option value="date">Date</option>
            <option value="amount">Amount</option>
          </select>
        </div>

        <br/>
        
        <DateRangePicker
          startDate={this.props.RootStore.FiltersStore.filters.startDate}
          endDate={this.props.RootStore.FiltersStore.filters.endDate}
          onDatesChange={this.onDatesChange}
          focusedInput={this.state.calendarFocused}
          onFocusChange={this.onFocusChange}
          showClearDates={true}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />

      </div>
    )
  } 
}


/* Without decorators */
// const ExpenseListFilters = inject('RootStore', 'FiltersAction')(observer((props)=> (
//   <div>
//     <input type="text" value={props.RootStore.FiltersStore.filters.text} onChange={(e) => {
//       props.FiltersAction.setTextFilter(e.target.value)
//     }}/>
//   </div>
// )))


export default ExpenseListFilters
