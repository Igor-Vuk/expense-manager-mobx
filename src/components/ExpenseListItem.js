import React, { Component } from 'react'
import { Link } from 'react-router-dom'

/* With decorators */
class ExpenseListItem extends Component {
  render() {
    const {description, amount, createdAt, id} = this.props
    return (
      <div>
        <Link to = {`/edit/${id}`}><h3>{description}</h3></Link>
        <p>{amount} - {createdAt}</p>
      </div>
    )
  }
}

/* Without decorators */
// const ExpenseListItem = (({...props, description, amount, createdAt, id}) => (
//   <div>
//     <h3>{description}</h3>
//     <p>{amount} - {createdAt}</p>
//   </div>
// ))

export default ExpenseListItem
