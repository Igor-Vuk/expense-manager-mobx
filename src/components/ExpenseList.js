import React, {Component} from 'react'
import { observer, inject } from 'mobx-react'
import ExpenseListItem from './ExpenseListItem'
import selectExpense from '../selectors/expenses'

/* Using decorators */
@inject('RootStore')
@observer
export class ExpenseList extends Component {
  render() {
    const {expenses} = this.props.RootStore.ExpensesStore
    const {filters} = this.props.RootStore.FiltersStore
    const expensesFilter = selectExpense(expenses, filters)
    return (
      <div>
      {
        expenses.length === 0 ? (
          <p>No expenses</p>
        ) : (
          expensesFilter.map((expense) => {
            return <ExpenseListItem key={expense.id} {...expense} />
          })
        )
      }
      </div>
    )
  }
}

/* Without decorators */
// const ExpenseList = inject('RootStore')(observer(class ExpenseList extends Component {
//   render() {
//     const {expenses} = this.props.RootStore.ExpensesStore
//     const {filters} = this.props.RootStore.FiltersStore
//     const expensesFilter = selectExpense(expenses, filters)
//     return (
//       <div>
//         {expensesFilter.map((expense) => {
//           return <ExpenseListItem key={expense.id} {...expense} />
//         })}
//       </div>
//     )
//   }
// }))

export default ExpenseList