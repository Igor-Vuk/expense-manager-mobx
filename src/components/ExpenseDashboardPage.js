import React from 'react'
import ExpenseList from './ExpenseList'
import ExpenseListFilters from './ExpenseListFilters'
// import DevTools from 'mobx-react-devtools';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList/>
  </div>
)

export default ExpenseDashboardPage