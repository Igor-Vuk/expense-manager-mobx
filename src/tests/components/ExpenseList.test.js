import React from 'react'
import { shallow } from 'enzyme'
import { ExpenseList } from '../../components/ExpenseList'
import expenses from '../fixtures/expenses'
import moment from'moment'

/* Using shallow rendering won't provide any useful results when testing injected components; only the injector will be rendered. To test with shallow rendering, instantiate the wrappedComponent */

test('should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList.wrappedComponent 
  RootStore = {
    {
      ExpensesStore: {
        expenses
      },
      FiltersStore: {
        filters: {
          text: 'e',
          sortBy: 'date',
          startDate: '',
          endDate: ''
        }
      }
    }
  } 
  />)
  expect(wrapper).toMatchSnapshot();
})
