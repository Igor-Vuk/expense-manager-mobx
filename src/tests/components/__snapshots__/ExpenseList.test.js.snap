// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList with expenses 1`] = `
<div>
  <ExpenseListItem
    amount={4500}
    createdAt={345600000}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
  <ExpenseListItem
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
</div>
`;
